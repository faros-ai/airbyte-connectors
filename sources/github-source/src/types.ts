export interface GithubConfig {
  authentication: GitHubAuth;
  api_url?: string;
  concurrency_limit?: number;
  reject_unauthorized?: boolean;
  previews?: ReadonlyArray<string>;
}

type GitHubToken = {
  auth: 'token';
  personal_access_token: string;
};

type PartialGitHubApp = {
  auth: 'app';
  app_id: number;
  private_key: string;
};

type GitHubAppInstallation = PartialGitHubApp & {
  app_cfg: {
    auth: 'installation';
    installation_id: number;
  };
};

type GitHubAppClient = PartialGitHubApp & {
  app_cfg: {
    auth: 'client';
    client_id: string;
    client_secret: string;
  };
};

export type GitHubAuth = GitHubToken | GitHubAppClient | GitHubAppInstallation;
