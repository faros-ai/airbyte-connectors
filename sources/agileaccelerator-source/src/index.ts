import {Command} from 'commander';
import {
  AirbyteLogger,
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {
  Agileaccelerator,
  AgileacceleratorConfig,
} from './agileaccelerator/agileaccelerator';
import {Works} from './streams';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteLogger();
  const source = new AgileacceleratorSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** AgileAccelerator source implementation. */
export class AgileacceleratorSource extends AirbyteSourceBase<AgileacceleratorConfig> {
  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(
    config: AgileacceleratorConfig
  ): Promise<[boolean, VError]> {
    try {
      const agileaccelerator = await Agileaccelerator.instance(
        config,
        this.logger
      );
      await agileaccelerator.checkConnection(config);
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AgileacceleratorConfig): AirbyteStreamBase[] {
    return [new Works(config, this.logger)];
  }
}
