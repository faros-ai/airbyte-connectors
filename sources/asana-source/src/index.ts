import {Command} from 'commander';
import {
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import {AirbyteSourceLogger} from 'faros-airbyte-cdk';
import VError from 'verror';

import {Asana, AsanaConfig} from './asana';
import {
  Projects,
  ProjectTasks,
  Tags,
  Tasks,
  Users,
  Workspaces,
} from './streams';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteSourceLogger();
  const source = new AsanaSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** Asana source implementation. */
export class AsanaSource extends AirbyteSourceBase<AsanaConfig> {
  get type(): string {
    return 'asana';
  }

  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(config: AsanaConfig): Promise<[boolean, VError]> {
    try {
      const asana = Asana.instance(config, this.logger);
      await asana.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AsanaConfig): AirbyteStreamBase[] {
    return [
      new Projects(config, this.logger),
      new Tags(config, this.logger),
      new Tasks(config, this.logger),
      new Users(config, this.logger),
      new Workspaces(config, this.logger),
      new ProjectTasks(config, this.logger),
    ];
  }
}
