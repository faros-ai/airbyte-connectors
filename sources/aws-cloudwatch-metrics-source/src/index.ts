import {Command} from 'commander';
import {
  AirbyteSourceBase,
  AirbyteSourceLogger,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {CloudWatch, Config} from './cloudwatch';
import {Metrics} from './streams';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteSourceLogger();
  const source = new CloudWatchMetricsSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** CloudWatch Metrics source implementation. */
export class CloudWatchMetricsSource extends AirbyteSourceBase<Config> {
  get type(): string {
    return 'cloudwatch-metrics';
  }

  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(config: Config): Promise<[boolean, VError]> {
    try {
      const cloudWatch = CloudWatch.instance(config);
      await cloudWatch.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: Config): AirbyteStreamBase[] {
    return [new Metrics(config, this.logger)];
  }
}
