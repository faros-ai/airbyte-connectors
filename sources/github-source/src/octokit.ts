import {createAppAuth} from '@octokit/auth-app';
import {retry} from '@octokit/plugin-retry';
import {throttling, ThrottlingOptions} from '@octokit/plugin-throttling';
import {Octokit} from '@octokit/rest';
import Bottleneck from 'bottleneck';
import {AirbyteLogger, AirbyteLogLevel} from 'faros-airbyte-cdk';
import {getOperationAST, parse} from 'graphql';
import https from 'https';
import {uniq} from 'lodash';
import {Dictionary} from 'ts-essentials';
import url from 'url';
import template from 'url-template';
import VError from 'verror';

import {GithubConfig} from './types';

export type RateLimitHandler = (reason?: Error) => boolean;

const ExtendedOctokit = Octokit.plugin(retry, throttling);

export const GITHUB_API_URL = 'https://api.github.com';

export function makeOctokitClient(
  cfg: GithubConfig,
  logger?: AirbyteLogger,
  onRateLimit?: RateLimitHandler,
  maxRetries = 3
): Octokit {
  const throttle = getThrottle(cfg, logger, onRateLimit, maxRetries);
  const baseUrl = cfg.api_url ?? GITHUB_API_URL;
  // Check whether the protocol matches 'https:'
  const isHttps = new url.URL(baseUrl).protocol.startsWith('https');
  const request = isHttps
    ? {
        agent: new https.Agent({
          rejectUnauthorized: cfg.reject_unauthorized ?? true,
        }),
        timeout: 90_000,
      }
    : {timeout: 90_000};

  const auth = getOctokitAuth(cfg);

  const kit = new ExtendedOctokit({
    auth,
    authStrategy: cfg.authentication.auth === 'app' ? createAppAuth : undefined,
    baseUrl,
    request,
    throttle,
  });

  kit.hook.before('request', (request) => {
    beforeRequestHook(cfg, request, logger);
  });

  kit.hook.after('request', (response) => {
    logger.debug(
      `Response: ${response.status} ${response.url} [remaining ${response.headers['x-ratelimit-resource']} rate-limit quota: ${response.headers['x-ratelimit-remaining']}]`
    );
  });

  return kit;
}

function getOctokitAuth(cfg: GithubConfig): string | Dictionary<any> {
  if (cfg.authentication?.auth === 'token') {
    if (!cfg.authentication.personal_access_token) {
      throw new VError(
        'Invalid token configuration: Personal access token is required'
      );
    }
    return cfg.authentication.personal_access_token;
  }

  if (cfg.authentication?.auth === 'app') {
    if (
      !cfg.authentication.app_id ||
      !cfg.authentication.private_key ||
      !cfg.authentication.app_cfg
    ) {
      throw new VError(
        'Invalid app configuration: app_id, private_key and app_cfg are required'
      );
    }

    if (cfg.authentication.app_cfg.auth === 'client') {
      if (
        !cfg.authentication.app_cfg.client_id ||
        !cfg.authentication.app_cfg.client_secret
      ) {
        throw new VError(
          'Invalid app client configuration: client_id and client_secret are required'
        );
      }
      return {
        type: cfg.authentication.auth,
        appId: cfg.authentication.app_id,
        privateKey: cfg.authentication.private_key,
        clientId: cfg.authentication.app_cfg.client_id,
        clientSecret: cfg.authentication.app_cfg.client_secret,
      };
    }

    if (cfg.authentication.app_cfg.auth === 'installation') {
      if (!cfg.authentication.app_cfg.installation_id) {
        throw new VError(
          'Invalid app installation configuration: installation_id is required'
        );
      }
      return {
        type: cfg.authentication.auth,
        appId: cfg.authentication.app_id,
        privateKey: cfg.authentication.private_key,
        installationId: cfg.authentication.app_cfg.installation_id,
      };
    }

    throw new VError('Invalid value for authentication.app_cfg.auth');
  }

  throw new VError('Invalid value for authentication.auth');
}

function getThrottle(
  cfg: GithubConfig,
  logger: AirbyteLogger,
  onRateLimit: (reason?: Error) => boolean,
  maxRetries: number
): ThrottlingOptions & {global: Bottleneck.Group} {
  return {
    global: new Bottleneck.Group({
      minTime: 100,
      maxConcurrent: cfg.concurrency_limit,
    }),
    onRateLimit: (after: number, opts: any): boolean | undefined => {
      const msg = `Quota exhausted for ${opts.method} ${opts.url}`;
      logger.warn(msg);
      return onRateLimit(new Error(msg));
    },
    onSecondaryRateLimit: (after: number, opts: any): boolean | undefined => {
      logger.warn(
        `Abuse detected for ${opts.method} ${opts.url}. Retry count: ${opts.request.retryCount}, after: ${after}`
      );
      if (opts.request.retryCount < maxRetries) {
        logger.info(`Retrying after ${after} seconds.`);
        return true;
      }
    },
  };
}

function beforeRequestHook(
  cfg: GithubConfig,
  request: any,
  logger: AirbyteLogger
): void {
  // Allow setting media type for preview features
  // E.g https://docs.github.com/en/enterprise-server@3.1/rest/overview/api-previews
  if (cfg.previews?.length) {
    const mediaType = request.mediaType;
    request.mediaType = {
      ...mediaType,
      previews: uniq((mediaType?.previews ?? []).concat(cfg.previews)),
    };
  }
  if (logger.level === AirbyteLogLevel.DEBUG) {
    let url = request.url;
    if (url.includes('{')) {
      const urlTemplate = template.parse(url);
      url = urlTemplate.expand(request);
    }
    if (url.startsWith('/')) {
      url = request.baseUrl + url;
    }

    let query = '';
    if (request?.query && url.endsWith('/graphql')) {
      const queryStr = String(request.query);
      const node = parse(queryStr);
      const operation = getOperationAST(node);
      const vars = request?.variables ?? {};
      query = `[${operation.name.value}: ${JSON.stringify(vars)}]`;
    }

    logger.debug(`Request : ${request.method} ${url} ${query}`);
  }
}
