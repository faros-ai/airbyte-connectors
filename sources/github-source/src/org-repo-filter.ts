import {AirbyteLogger} from 'faros-airbyte-cdk';
import {FarosClient} from 'faros-js-client';
import {toLower} from 'lodash';
import {Memoize} from 'typescript-memoize';

import {GitHub} from './github';
import {GitHubConfig} from './types';

export class OrgRepoFilter {
  organizations: Set<string> | undefined;

  constructor(
    private readonly config: GitHubConfig,
    private readonly logger: AirbyteLogger,
    private readonly farosClient?: FarosClient
  ) {}

  @Memoize()
  async getOrganizations(): Promise<ReadonlyArray<string>> {
    if (!this.organizations) {
      this.organizations = new Set();
      const github = await GitHub.instance(this.config, this.logger);
      if (!this.config.organizations) {
        const orgs = await github.getOrganizations();
        for await (const org of orgs) {
          this.organizations.add(org);
        }
      } else {
        for (const org of this.config.organizations) {
          this.organizations.add(toLower(org));
        }
      }
    }
    return Array.from(this.organizations);
  }
}
