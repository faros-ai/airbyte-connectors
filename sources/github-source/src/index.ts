import {Command} from 'commander';
import {
  AirbyteSourceBase,
  AirbyteSourceLogger,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {makeOctokitClient} from './octokit';
import {GithubConfig} from './types';

export function mainCommand(): Command {
  const logger = new AirbyteSourceLogger();
  const source = new GithubSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

export class GithubSource extends AirbyteSourceBase<GithubConfig> {
  get type(): string {
    return 'github';
  }

  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(config: GithubConfig): Promise<[boolean, VError]> {
    try {
      const octokit = makeOctokitClient(config, this.logger);
      await octokit.users.getAuthenticated();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: GithubConfig): AirbyteStreamBase[] {
    return [];
  }
}
