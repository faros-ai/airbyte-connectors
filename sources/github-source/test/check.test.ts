import {
  AirbyteLogLevel,
  AirbyteSourceLogger,
  sourceCheckTest,
} from 'faros-airbyte-cdk';

import * as sut from '../src/index';

describe('index', () => {
  const logger = new AirbyteSourceLogger(
    // Shush messages in tests, unless in debug
    process.env.LOG_LEVEL === 'debug'
      ? AirbyteLogLevel.DEBUG
      : AirbyteLogLevel.FATAL
  );
  const source = new sut.GithubSource(logger);

  async function runSourceCheckTest(configOrPath: string): Promise<void> {
    return sourceCheckTest({
      source,
      configOrPath,
    });
  }

  test('check connection - token valid', async () => {
    // TODO: mock API call
    await runSourceCheckTest('check_connection/token_valid.json');
  });

  test('check connection - token missing', async () => {
    await runSourceCheckTest('check_connection/token_invalid.json');
  });

  test('check connection - app invalid', async () => {
    await runSourceCheckTest('check_connection/app_invalid.json');
  });

  test('check connection - app auth missing', async () => {
    await runSourceCheckTest('check_connection/app_auth_missing.json');
  });

  test('check connection - app client invalid', async () => {
    await runSourceCheckTest('check_connection/app_client_invalid.json');
  });

  test('check connection - app client valid', async () => {
    // TODO: mock API call
    await runSourceCheckTest('check_connection/app_client_valid.json');
  });

  test('check connection - app installation invalid', async () => {
    await runSourceCheckTest('check_connection/app_installation_invalid.json');
  });

  test('check connection - app installation valid', async () => {
    // TODO: mock API call
    await runSourceCheckTest('check_connection/app_installation_valid.json');
  });

  test('check connection - authentication missing', async () => {
    await runSourceCheckTest('check_connection/authentication_missing.json');
  });
});
