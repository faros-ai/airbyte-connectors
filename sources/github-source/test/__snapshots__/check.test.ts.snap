// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index check connection - app auth missing 1`] = `
AirbyteConnectionStatusMessage {
  "connectionStatus": {
    "message": "Invalid value for authentication.app_cfg.auth",
    "status": "FAILED",
  },
  "type": "CONNECTION_STATUS",
}
`;

exports[`index check connection - app client invalid 1`] = `
AirbyteConnectionStatusMessage {
  "connectionStatus": {
    "message": "Invalid app client configuration: client_id and client_secret are required",
    "status": "FAILED",
  },
  "type": "CONNECTION_STATUS",
}
`;

exports[`index check connection - app client valid 1`] = `
AirbyteConnectionStatusMessage {
  "connectionStatus": {
    "message": "[@octokit/auth-app] installationId option is required for installation authentication.",
    "status": "FAILED",
  },
  "type": "CONNECTION_STATUS",
}
`;

exports[`index check connection - app installation invalid 1`] = `
AirbyteConnectionStatusMessage {
  "connectionStatus": {
    "message": "Invalid app installation configuration: installation_id is required",
    "status": "FAILED",
  },
  "type": "CONNECTION_STATUS",
}
`;

exports[`index check connection - app installation valid 1`] = `
AirbyteConnectionStatusMessage {
  "connectionStatus": {
    "message": "[@octokit/auth-app] installationId option is required for installation authentication.",
    "status": "FAILED",
  },
  "type": "CONNECTION_STATUS",
}
`;

exports[`index check connection - app invalid 1`] = `
AirbyteConnectionStatusMessage {
  "connectionStatus": {
    "message": "Invalid app configuration: app_id, private_key and app_cfg are required",
    "status": "FAILED",
  },
  "type": "CONNECTION_STATUS",
}
`;

exports[`index check connection - authentication missing 1`] = `
AirbyteConnectionStatusMessage {
  "connectionStatus": {
    "message": "Invalid value for authentication.auth",
    "status": "FAILED",
  },
  "type": "CONNECTION_STATUS",
}
`;

exports[`index check connection - token missing 1`] = `
AirbyteConnectionStatusMessage {
  "connectionStatus": {
    "message": "Invalid token configuration: Personal access token is required",
    "status": "FAILED",
  },
  "type": "CONNECTION_STATUS",
}
`;

exports[`index check connection - token valid 1`] = `
AirbyteConnectionStatusMessage {
  "connectionStatus": {
    "message": "Bad credentials - https://docs.github.com/rest",
    "status": "FAILED",
  },
  "type": "CONNECTION_STATUS",
}
`;
